## Introduction: 

This notebook is devoted to the 3rd question in Chapter-1, regarding Pregnencies data exploration .. 

The website for the textbook provides code to process the data (called Survey.py) and we gonna use it in this exercice .. 

http://greenteapress.com/thinkstats/survey.py


## Ex. 3.1 :

### Printing the number of records .. 

In this exercise you will write a program to explore the sata in the Pregnancies table. 

1. In the directory where you put survey.py and the data files, create file named first.py and type or paste in the following code and check the results: 

        import survey

        table = survey.Pregnancies()

        table.ReadRecords()

        print 'Number of pregnancies: ', len(table.records)

In [68]:
import survey
import first
import numpy as np
import pandas as pd 

In [69]:
table = survey.Pregnancies()

In [70]:
table.ReadRecords()

In [71]:
print 'Number of pregnancies: ', len(table.records)

Number of pregnancies:  13593


## Justification: 

These statements above print the number of records in the pregnancies file .. the result matches the output mentioned in the textbook .. 

## Ex 3.2 :


### count alive .. 

Write a loop that iterates table and counts the number of live births. 

Find the documentation of outcome and confirm that your result is consistent with the summary in the documentation.

In [72]:
data = table.GetFields()

In [73]:
data

[('caseid', 1, 12, int),
 ('nbrnaliv', 22, 22, int),
 ('babysex', 56, 56, int),
 ('birthwgt_lb', 57, 58, int),
 ('birthwgt_oz', 59, 60, int),
 ('prglength', 275, 276, int),
 ('outcome', 277, 277, int),
 ('birthord', 278, 279, int),
 ('agepreg', 284, 287, int),
 ('finalwgt', 423, 440, float)]

In [74]:
## extract outcome column .. 

outcome_ = np.zeros([len(table.records), 1])

for i in xrange(len(table.records)):
    outcome_[i] = table.records[i].outcome

In [75]:
outcome_

array([[ 1.],
       [ 1.],
       [ 1.],
       ..., 
       [ 2.],
       [ 1.],
       [ 1.]])

In [76]:
alive = 0

for baby in outcome_:
    if (baby == 1):
        alive = alive + 1
    else:
        continue 

In [77]:
alive

9148

## Justification: 

In the documentation they menioned the folowing: 

    OUTCOME = Outcome of pregnancy (=1 if live birth) 

The above result is consistant with the documention results (that output 9148 alive babies) .. 

The documenantion results link ==> 

http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=&subSec=8014&srtLabel=611785

## Ex 3.3 :


### Partition the data .. 

Modify the loop to partition the live birth records into two groups, one for first babies and one for the others. 

Again, read the documentation of birthord to see if your results are consistent.

==> First of all, I tried to read the file in lines and push these lines in array, convert the array to DataFrame and then partition the dataframe table into two smaller tables (one for first babies and the second one for Others) ..  

In [78]:
## read all data in dataframe .. then partition the dataframe table into two tables 
df = pd.DataFrame([])

header =  [case[0] for (i, case) in enumerate(data)]  ## append, appends rows !! 
header = np.transpose(header)
header

array(['caseid', 'nbrnaliv', 'babysex', 'birthwgt_lb', 'birthwgt_oz',
       'prglength', 'outcome', 'birthord', 'agepreg', 'finalwgt'], 
      dtype='|S11')

In [ ]:
Value_to_push = []
with open("2002FemPreg.dat","r") as io:
    for l in io:
        if len(l) > 100:
            for (i,case) in enumerate(data):
                try: 
                    Value_to_push.append([case[3](l[case[1]:case[2]])])
                    print(Value_to_push)
                except ValueError: 
                    continue

In [27]:
# alive_partition = []
# others_partition = []

# for record in xrange(len(table.records)):
#     if(table.records[record].outcome == 1):
#         alive_partition.append(table.records[record])
#     else:
#         others_partition.append(table.records[record])

# print(np.shape(alive_partition), np.shape(others_partition))

## Comment: 

the above trail didn't work properly as append(), adds values at the end of the list .. this is help with constructing tabuler form (each new entry is seen as single value) .. 

I tried insert() function .. but it inserted the whole records at the end of the end of the array (each new entry is seen as record) .. 

Finally, I used the first.py class mentioned in the textbook, utilizing its built in function (PartitionRecords) in partitioning the data


In [79]:
partitions = first.PartitionRecords(table)
partitions

(<survey.Pregnancies at 0x7f6d32035f50>,
 <survey.Pregnancies at 0x7f6d32035f90>)

In [80]:
first_baby = partitions[0]
others = partitions[1]
len(first_baby), len(others)

(4413, 4735)

## Justification:

In the documentation they mentioned that the number of first babies are: 4413 which matches the results above ..

They implicitly mentioned the number of (others) babies, by listing ALL possible number of births + inapplicable values (i.e. NAs) .. by subtracting the NAs from the total number of records  we gonna obtain the number of (others) 

Others length = 13593 - 4445 ==> 4735

http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=&subSec=8016&srtLabel=611933

## Note:
We still can apply the first way of thinking of dividing the big table into two smaller ones using the built in function (MakeTables) in first.py 

http://greenteapress.com/thinkstats/first.py


In [95]:
t = first.MakeTables()
original_table = t[0]
firsts_table = t[1]
others_table = t[2]

original_table, firsts_table, others_table

(<survey.Pregnancies at 0x7f6d30c6d090>,
 <survey.Pregnancies at 0x7f6d302cab50>)

In [98]:
len(original_table), len(firsts_table), len(others_table) ## same results 

(13593, 4413, 4735)

## Ex 3.4: 

Compute the average pregnancy length (in weeks) for first babies and others. Is there a difference between the groups? How big is it?

In [82]:
first_baby.GetFields()

[('caseid', 1, 12, int),
 ('nbrnaliv', 22, 22, int),
 ('babysex', 56, 56, int),
 ('birthwgt_lb', 57, 58, int),
 ('birthwgt_oz', 59, 60, int),
 ('prglength', 275, 276, int),
 ('outcome', 277, 277, int),
 ('birthord', 278, 279, int),
 ('agepreg', 284, 287, int),
 ('finalwgt', 423, 440, float)]

In [87]:
first_baby.GetFields()[5] ## the preglength column data related only to first_baby 

('prglength', 275, 276, int)

In [101]:
first.Summarize('/home/eman/Dr.Beltagy Pre_Phd/Session 5 - diff notebooks')

Number of first babies 4413
Number of others 4735
Mean gestation in weeks:
First babies 38.6009517335
Others 38.5229144667
Difference in days 0.546260867443


## Comment:
The average gestation (in weeks) between the 1st babies and others is about 0.08 which means it is not that big ! .. 

This is an evidance that 1st babies don't necessarly come late !! (the core question of the textbook)  

## Justification:

There is no evidance for the correctness of the above results as they didn't mention the average in the documentation ..  only number of cases corresponding to each common preg_length .. 

http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=&subSec=8016&srtLabel=611931

Hence, I can assure the results by comparing them with the Julia (end to end) results .. 